Static deploy flags

denisristic opened the pull request:
When calling setup:static-content:deploy now you are able to put flags to skip specific content and you are able to do that for specific theme(s)

Usage:
setup:static-content:deploy [-d|--dry-run] [--no-javascript] [--no-css] [--no-less] [--no-images] [--no-fonts] [--no-html] [--no-misc] [--no-html-minify] [-t|--theme[="..."]] [-et|--exclude-theme[="..."]] [-l|--language[="..."]] [-el|--exclude-language[="..."]] [-a|--area[="..."]] [-ea|--exclude-area[="..."]] [languages1] ... [languagesN]

Options:
 --dry-run (-d)      If specified, then no files will be actually deployed.
 --no-javascript    If specified, no JavaScript will be deployed.
 --no-css              If specified, no CSS will be deployed.
 --no-less             If specified, no LESS will be deployed.
 --no-images        If specified, no images will be deployed.
 --no-fonts            If specified, no font files will be deployed.
 --no-html             If specified, no html files will be deployed.
 --no-misc             If specified, no miscellaneous files will be deployed.
 --no-html-minify   If specified, just html will not be minified and actually deployed.
 --theme (-t)         If specified, just specific themes will be deployed (multiple values allowed)
 --exclude-theme (-et)  If specified, specific theme will not be deployed (multiple values allowed)
 --language (-l)     If specified, just specific language will be deployed (multiple values allowed)
 --exclude-language (-el) If specified, specific language will not be deployed (multiple values allowed)
 --area (-a)           If specified, just specific area will be deployed (multiple values allowed)
 --exclude-area (-ea)  If specified, specific area will not be deployed (multiple values allowed)

Commit:
Added flags to static deploy command to skip filetypes (js, css, less, images, fonts...) and filter by theme

Commit:
Added flags to static deploy command to skip filetypes (js, css, less, images, fonts...) and filter by theme

Commit:
Revert "Added flags to static deploy command to skip filetypes (js, css, less, images, fonts...) and filter by theme"

This reverts commit 4e6548620f07fbbaa121f94ecd79d6b9d400b7c1.

Commit:
Revert "Added flags to static deploy command to skip filetypes (js, css, less, images, fonts...) and filter by theme"

This reverts commit 4e6548620f07fbbaa121f94ecd79d6b9d400b7c1.

Commit:
Revert "Revert "Added flags to static deploy command to skip filetypes (js, css, less, images, fonts...) and filter by theme""

This reverts commit 04ebbe796c2a11608581fecdf4d9b6645e19d2e1.

Commit:
Revert "Revert "Revert "Added flags to static deploy command to skip filetypes (js, css, less, images, fonts...) and filter by theme"""

This reverts commit 8216e1253cf4e42cd1c375392f0948c4e80c1410.

Commit:
Added flags to static deploy command to skip filetypes (js, css, less, images, fonts...) and filter by theme - fixed

adragus-inviqa commented on app/code/Magento/Deploy/Model/Deployer.php:
```diff
@@ -105,20 +152,59 @@ function (AlternativeSourceInterface $alternative) {
+                return true;
+            } else  if ($this->isFonts && in_array($ext, array('eot', 'svg', 'ttf', 'woff', 'woff2'))) {
+                return true;
+            } else  if ($this->isMisc && in_array($ext, array('md', 'txt', 'jbf', 'csv', 'json', 'txt', 'htc', 'swf', 'LICENSE', ''))) {
```
Please use `elseif` instead of `else if`, or `else  if` "so that all control keywords look like single words".

adragus-inviqa replied:
I would rather you put those file extension strings somewhere - if not configurable - at least more visible, like some private properties.

denisristic replied:
Done, thanks for your remarks

adragus-inviqa commented on app/code/Magento/Deploy/Model/Deployer.php:
```diff
@@ -105,20 +152,59 @@ function (AlternativeSourceInterface $alternative) {
+     * @SuppressWarnings(PHPMD.CyclomaticComplexity)
+     * @SuppressWarnings(PHPMD.NPathComplexity)
+     */
+    private function checkSkip($filePath) {
```
`{` on its own line.

adragus-inviqa commented:
Please use PHP 5.4's array notation - `[]` - instead of `array`. It's less noisy that way.

Commit:
Fixed code style, remarks by @adragus-inviqa

Commit:
Fixed method parameter instatiation when paramter is empty

mazhalai commented:
@denisristic thank you for your contribution, we have created MAGETWO-52614 to process this PR.

piotrekkaminski commented:
@denisristic can you sign the contributor license agreement? it is stopping us from merging this nice improvement.

denisristic commented:
@piotrekkaminski I've signed it several times but back then it didn't change status. As I can see now it's changed even though I haven't done anything

hostep commented:
@denisristic: I happen to notice you changed the file permissions of the two php files from 644 to 755, I think you'll need to put them back to 644.

Other than that, @ Magento team: do you think this PR will be included in Magento version 2.1?
I haven't tested it yet, but it looks like the `--themes` switch will allow us to significantly improve the time for each deployment on a production server.

hostep commented:
Just did some testing using this patch, in case anyone is interested (I'm running in production mode btw):

``` sh
# First remove all caches
$ rm -R pub/static/*
$ rm -R var/cache/*
$ rm -R var/page_cache/*
$ rm -R var/generation/*
$ rm -R var/view_preprocessed/*

# Now run static deploy without a filter
$ time php bin/magento setup:static-content:deploy

...

New version of deployed files: 1464678133

real    6m35.824s
user    6m24.147s
sys     0m14.243s


# Now do the same, first removing all caches again
$ rm -R ...

# Run static deploy with filter on backend theme and our custom theme
$ time php bin/magento setup:static-content:deploy --themes Magento/backend --themes Vendor/custom

...

New version of deployed files: 1464678974

real    2m32.092s
user    2m23.732s
sys     0m5.688s

```

Over 2 times faster, wow!

Disclaimer: I have 2 custom themes set up, so the first run should be faster if you only have 1 custom theme.

Thanks @denisristic, awesome work!

Hope this gets merged in Magento 2.1 codebase.

erikhansen commented:
@denisristic Thanks for working on this PR. Two suggestions/requests:
1. You may want to rename your `--themes` flag to `--theme` since it looks like you have to pass separate arguments for each theme you want to include. Alternatively you could rework your `--themes` flag to accept a comma-delimited list of themes. For example: `bin/magento setup:static-content:deploy --themes=vendor/theme1,vendor/theme2`
2. Per [my comment](https://github.com/magento/magento2/issues/3038#issuecomment-192007804) on a issue #3038, it would be great if you could add an `--exclude-theme` flag that would allow a developer to specify which themes NOT to include. This would be especially useful for a Magento site where new themes might be added over time and the merchant is using deployment scripts that have hard-coded deployment steps. For example: `bin/magento setup:static-content:deploy --exclude-theme=magento/luma --exclude-theme=magento/blank`

erikhansen commented:
@denisristic Are you interested in adding a `--exclude-theme` flag as I suggested in my previous comment? If not, let me know.

@piotrekkaminski Can you confirm that you're intending to merge this PR once the current merge conflicts are resolved and it goes through code review?

denisristic commented:
@erikhansen Thanks for your comments. I`ll add exclude flag version for themes until the end of the week.

piotrekkaminski commented:
@erikhansen yes - it is very high in the backlog for next version (2.2), however won't be possible for 2.1.

giacmir commented:
It would be nice to have a switch to automatically grab all configured languages without need to explicitly pass them as arguments. In case of installations with several languages it is pretty easy to forget one, and it is very useful for automated deploys.

erikhansen commented:
@denisristic Any update on adding the `--exclude-theme` flag and resolving the merge conflicts in this branch so that this PR can be merged in?

Commit:
RENAMED --themes to -theme (@erikhansen suggestion)
ADDDED --exclude-theme flag (@erikhansen suggestion)
UPDATED --language flag (@giacmir)
ADDED flags: --area (-a), --exclude-area (-ea), --exclude-language (-el), --exclude-theme (-et)
ADDED flag shotcuts: -l, -el, -t, -et, -a, -ea
ADDED check for wrong language, theme and/or area parameter value
FIXED variable names, changet do camelCase

erikhansen commented:
@denisristic Thanks for adding all of the suggested changes! It looks like there are still merge conflicts. Could you merge develop into your branch and resolve those merge conflicts?

Commit:
Merge remote-tracking branch 'origin/develop' into static-deploy-flags

merge with 2.1 codbase

Commit:
merged with 2.1 codebase
changed permissions as @hostep suggested

Commit:
Merge branch 'develop' into static-deploy-flags

# Conflicts:
#	app/code/Magento/Deploy/Console/Command/DeployStaticContentCommand.php
#	app/code/Magento/Deploy/Model/Deployer.php

denisristic commented:
@erikhansen I've resolved merge conflicts. Waiting for travis CI to check if everything is OK

Commit:
FIXED merge confilct

Commit:
FIXED Travis CI errors (PHPMD, PHPCS...)

Commit:
FIXED Avoid unused private fields such as '$fileExtensionsHtml'.
FIXED Avoid unused private methods such as 'checkLanguagesInput'
FIXED Avoid unused parameters such as '$locales'.
FIXED restored original languages argument

Commit:
FIXED Avoid unused local variables such as '$languagedInclude'.
FIXED Invalid argument supplied for foreach()
FIXED code style (PHPCS & PHPMD)

mslabko commented:
Hi @denisristic 
Could you please sign the License Agreement? https://cla.dev.magento.com/magento/magento2?pullRequest=4294

denisristic commented:
Hi @mslabko ,
I've done it at least 20 times but it didn't change.
I can see that it is signed now even though I didn't do anything in last 10 days

mslabko commented:
@denisristic ,
Thank you! Your PR already in progress.

sshrewz commented:
@denisristic, we need you to sign the CLA before we can proceed with further actions your contribution.

adragus-inviqa commented:
> @denisristic, we need you to sign the CLA before we can proceed with further actions your contribution.

Not only did @denisristic say that he signed the thing 10 times, but I'm seeing it signed already. Fix the signing process already.

denisristic commented:
@snohe  I've signed it at least 20x, there is some bug with signing process. As I can see now it is signed even though I didn't do anything

mslabko commented:
@snohe PR already in progress, CLA is applied.

mslabko commented:
@denisristic , PR was merged to mainline. Thank you for the contribution!
